src/mpid/ch4/netmod/ofi/ofi_nic.c

```diff
@@ -80,67 +79,83 @@ bool MPIDI_OFI_nic_already_used(const struct fi_info * prov, struct fi_info ** o
 
 /* Setup the multi-NIC data structures to use the fi_info structure given in prov */
 static int setup_single_nic(void);
-static int setup_multi_nic(int nic_count);
+static int setup_multi_nic(size_t nic_count);
 
 int MPIDI_OFI_init_multi_nic(struct fi_info *prov)
 {
     int mpi_errno = MPI_SUCCESS;
     int nic_count = 0;
-    int max_nics = MPIR_CVAR_CH4_OFI_MAX_NICS;
-
-    if (MPIR_CVAR_CH4_OFI_MAX_NICS == 0 || MPIR_CVAR_CH4_OFI_MAX_NICS <= -2) {
-        /* Invalid values for the CVAR will default to single nic */
-        max_nics = 1;
-    }
 
     /* Count the number of NICs */
-    struct fi_info *first_prov = NULL;
-    for (struct fi_info * p = prov; p; p = p->next) {
+    struct fi_info *first_prov = NULL, *prev_prov = NULL, *cur_prov = NULL;
+    for (cur_prov = prov; cur_prov; cur_prov = cur_prov->next) {
         /* additional filtering */
-        if (MPIR_CVAR_OFI_SKIP_IPV6 && p->addr_format == FI_SOCKADDR_IN6) {
+        if (MPIR_CVAR_OFI_SKIP_IPV6 && cur_prov->addr_format == FI_SOCKADDR_IN6) {
             continue;
         }
-        if (!MPIDI_OFI_nic_is_up(p)) {
+        if (!MPIDI_OFI_nic_is_up(cur_prov)) {
             continue;
         }
-        if (!first_prov) {
-            first_prov = p;
-        }
 #ifdef HAVE_LIBFABRIC_NIC
         /* check the nic */
-        struct fid_nic *nic = p->nic;
+        struct fid_nic *nic = cur_prov->nic;
         if (nic && nic->bus_attr->bus_type == FI_BUS_PCI &&
-            !MPIDI_OFI_nic_already_used(p, MPIDI_OFI_global.prov_use, nic_count)) {
-            MPIDI_OFI_global.prov_use[nic_count] = fi_dupinfo(p);
-            MPIR_Assert(MPIDI_OFI_global.prov_use[nic_count]);
-            nic_count++;
-            if (nic_count == max_nics) {
-                break;
+            !MPIDI_OFI_nic_already_used(cur_prov, first_prov)) {
+            if (prev_prov) {
+                prev_prov->next = fi_dupinfo(cur_prov);
+                prev_prov = prev_prov->next;
+            } else {
+                first_prov = fi_dupinfo(cur_prov);
+                prev_prov = first_prov;
             }
+            prev_prov->next = NULL;
+            ++nic_count;
         }
-#endif
+#else
+        /* Always pick the first provider if NIC info is not available */
+        if (!first_prov) {
+            first_prov = fi_dupinfo(cur_prov);
+            first_prov->next = NULL;
+            ++nic_count;
+            break;
+        }
+#endif /* HAVE_LIBFABRIC_NIC */
+    }
```

Comment:
This change can be extracted to a separate commit if you like. I kept it in the big commit since `git add -p` does not allow me to split it out.